### ISSUE TITLE
code implementation for handling dynamic queries

### ISSUE THREAD

**Issue by OPlincn** [hoiliu-0801/DQ-DETR#13]
I have gone through  your code but did not find the implementation for handling dynamic queries in a multi-batch setting. I noticed that in the DQ_5scale.py file, the batch_size is set to 1. Does this mean you train the model one image at a time to ensure that each image uses dynamic queries individually?

**Comment by RicePasteM** [comment]
I have been working with the code for several days. Based on my understanding, it appears that the number of queries is determined by the CCM module (located in `models/dqdetr/deformable_transformer.py`, line 330):
```python
counting_output, ccm_feature = self.CCM(memory, spatial_shapes)
multi_ccm_feature = self.multiscale(ccm_feature)
cgfe_out = self.CGFE(multi_ccm_feature, memory, spatial_shapes)
memory = cgfe_out        
_, predicted = torch.max(counting_output.data, 1)
num_select = self.dynamic_query_list[max(predicted.tolist())]
```
When using multiple batches, the number of queries will be the maximum value across all images in the batch for that epoch. For example, if Image 1 requires 1500 queries, Image 2 requires 900 queries, and Image 3 requires 300 queries, the final number of queries for the epoch will be 1500. This could potentially impact the performance of the CCM module.

Of course, this is just my interpretation, and I hope it helps. The original author might be able to provide a more accurate explanation.

**Comment by OPlincn** [comment]
> I have been working with the code for several days. Based on my understanding, it appears that the number of queries is determined by the CCM module (located in `models/dqdetr/deformable_transformer.py`, line 330):
> 
> counting_output, ccm_feature = self.CCM(memory, spatial_shapes)
> multi_ccm_feature = self.multiscale(ccm_feature)
> cgfe_out = self.CGFE(multi_ccm_feature, memory, spatial_shapes)
> memory = cgfe_out        
> _, predicted = torch.max(counting_output.data, 1)
> num_select = self.dynamic_query_list[max(predicted.tolist())]
> When using multiple batches, the number of queries will be the maximum value across all images in the batch for that epoch. For example, if Image 1 requires 1500 queries, Image 2 requires 900 queries, and Image 3 requires 300 queries, the final number of queries for the epoch will be 1500. This could potentially impact the performance of the CCM module.
> 
> Of course, this is just my interpretation, and I hope it helps. The original author might be able to provide a more accurate explanation.

Yes, I think so too. So I feel that the authors set the batch size to 1 to avoid using an unsuitable number of queries per image.

**Comment by Katie0723** [comment]
We set the batch size to 1 for training due to GPU memory limitations. In the current code, when using multiple batches, the number of queries is set to the maximum value across all images in the batch for that epoch. This setup "just" ensures that multiple batches can run without causing errors.

When using multiple batches, it is still possible to apply some attention mask so that each sample in the batch can have a different number of queries.

**Comment by OPlincn** [comment]
> We set the batch size to 1 for training due to GPU memory limitations. In the current code, when using multiple batches, the number of queries is set to the maximum value across all images in the batch for that epoch. This setup "just" ensures that multiple batches can run without causing errors.
> 
> When using multiple batches, it is still possible to apply some attention mask so that each sample in the batch can have a different number of queries.

okayyy Thank you very much!
